### PR TITLE
feat: implementation of shutdown operation

### PR DESCRIPTION
**Pull request by k-hayakawa-esol**
### Overview

This node accepts button operation input and controls shutdown. It also outputs to display the button operation status.

### Input and Output

- Input
  - from [button_output_selector](https://github.com/eve-autonomy/button_output_selector/)
    - `/shutdown_button` \[[autoware_state_machine_msgs/msg/VehicleButton][ReservationButton]\]:<br>Shutdown button input. (this topic is remapped from `input/shutdown_button`.)

- Output
  - to [delivery_reservation_lamp_manager](https://github.com/eve-autonomy/delivery_reservation_lamp_manager/)
    - `/shutdown_manager/state` \[[autoware_state_machine_msgs/msg/StateLock][StateLock]\]:<br>Request to change the lighting status for delivery reservation lamp. (this topic is remapped from `output/lock_state`.)

[StateLock]: https://github.com/eve-autonomy/autoware_state_machine_msgs/blob/main/msg/StateLock.msg
[ReservationButton]: https://github.com/eve-autonomy/autoware_state_machine_msgs/blob/main/msg/VehicleButton.msg

### Node Graph

node_graph

### Launch Parameter description

|ID|Name|Description|
|:---|:---|:----------|
|T0|time_required_to_release_button|Button hold down time[sec] required before shutdown is accepted.|
|T1|timeout_period_before_shutdown_aborts|Time[sec] given between accepting and executing shutdown (timeout after this time).|

### Shutdown state transition

state_transition

## Related Links

eve-autonomy/proj_launch#7

## Review Procedure

 * [x] Make sure the description is valid.
 * [x] Make sure that what is described in the description matches the implementation.

### CODE REVIEW

File: src/shutdown_manager.cpp [hunk 1]
```diff
@@ -0,0 +1,118 @@
...
+}
+
+void ShutdownManager::onTimer(void)
+{
+  switch (current_state_) {
+    case StateShutdown::STATE_INACTIVE_FOR_SHUTDOWN:
+      if (button_data_ && (button_hold_down_time >= time_required_to_release_button_)) {
+        time_to_start_pressing_button_ = this->now();
+        current_state_ = StateShutdown::STATE_STANDBY_FOR_SHUTDOWN;
+        publishShutdownState(current_state_);
```
Comment (yn-mrse) [inline src/shutdown_manager.cpp]: @k-hayakawa-esol 
Can't we use a specification that always outputs a state output regardless of its own state?
(Stopping the output under certain conditions would complicate the definition of interactions between nodes, so if there is no reason to do so, it is better not to do so.)

I want to eliminate all `publishShutdownState` functions and place only one between L110 and L111.

**Review by yn-mrse — CHANGES_REQUESTED** [review]

File: include/shutdown_manager/shutdown_manager.hpp [hunk 2]
```diff
@@ -0,0 +1,64 @@
...
+
+private:
+
+  uint16_t current_state_ = StateShutdown::STATE_INACTIVE_FOR_SHUTDOWN;
+  float time_required_to_release_button_ = 5.0;
+  float timeout_period_before_shutdown_aborts_ = 10.0;
+  builtin_interfaces::msg::Time time_to_start_pressing_button_;
+  bool button_data_ = false;
+  float button_hold_down_time = 0.0;
+  bool shutdown_executed_ = false;
```
Comment (yn-mrse) [inline include/shutdown_manager/shutdown_manager.hpp]: Please initialize member variables under the constructor.

**Review by yn-mrse — CHANGES_REQUESTED** [review]

File: src/shutdown_manager.cpp (hunk 1 above)
Comment (k-hayakawa-esol) [reply src/shutdown_manager.cpp]: Corrected timing of state publication.

File: include/shutdown_manager/shutdown_manager.hpp (hunk 2 above)
Comment (k-hayakawa-esol) [reply include/shutdown_manager/shutdown_manager.hpp]: Fixed to initialize in constructor.

**Review by yn-mrse — APPROVED** [review]

**Pull request merged**